core/domain/services/graphql/mutators/DatetimeDelete.php

```diff
@@ -35,7 +35,41 @@ public static function mutateAndGetPayload(EEM_Datetime $model, Datetime $type)
                 $entity = EntityMutator::getEntityFromInputData($model, $input);
 
                 // Delete the entity
-                $result = ! empty($input['deletePermanently']) ? $entity->delete_permanently() : $entity->delete();
+                if (! empty($input['deletePermanently'])) {
+                    // all related tickets
+                    $tickets = $entity->tickets();
+                    foreach ($tickets as $ticket) {
+                        // if the ticket is related to only one datetime
+                        if ($ticket->count_related('Datetime') === 1) {
+                            // Remove related prices for the ticket
+                            $ticket->delete_related_permanently('Price');
+                            // Remove relation with datetimes
+                            $ticket->_remove_relations('Datetime');
+                            // delete the ticket permanently
+                            $ticket->delete_permanently();
+                        }
+                    }
+
+                    // Remove relations with tickets
+                    $entity->_remove_relations('Ticket');
+                    // Now delete the datetime permanently
+                    $result = $entity->delete_permanently();
+                } else {
+                    // non trashed related tickets
+                    $tickets = $entity->tickets([[
+                        'TKT_deleted' => false,
+                    ]]);
+                    // loop though all tickets to check if we need to trash any
+                    foreach ($tickets as $ticket) {
+                        // if the ticket is related to a maximum of one datetime
+                        if ($ticket->count_related('Datetime') <= 1) {
+                            // trash the ticket
+                            $ticket->delete();
+                        }
+                    }
+                    // trash the datetime
+                    $result = $entity->delete();
+                }
```

Comment:
would be nicer / more readable to extract these blocks into separate functions like:

```php
 if (! empty($input['deletePermanently'])) {
	$result = DatetimeDelete::deleteDatetimeAndRelations($entity);
} else {
	$result = DatetimeDelete::trashDatetimeAndRelations($entity);
}
```

as well, the logic within the tickets loop is nearly identical to this code from `services/graphql/mutators/TicketDelete.php`

```php
// Remove related prices for the ticket
$entity->delete_related_permanently('Price');
// Remove relation with datetimes
$entity->_remove_relations('Datetime');
// Now delete the ticket permanently
$result = $entity->delete_permanently();
```

so that could be extracted into a public function and called from here.

```php
if ($ticket->count_related('Datetime') === 1) {
    TicketDelete::deleteTicketAndRelations($ticket);
}
```

Then if for some reason, this logic isn't quite right and needs to be updated, you only have to make a change in that one place:

Reply:
Done, thank you for the suggestion.

I don't why I don't feel confident about doing such things in PHP although I did think about doing that extraction 🙄